contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,212 @@
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52 weeks
+        for (uint i = 0; i < WEEKS_IN_YEAR; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
+                    /// @dev If no time change since last checkpoint just add new tokens
+                    /// that may have been deposited (same block)
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev In the event that toDistribute contains tokens
+                    /// for multiple weeks we take the remaining portion
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute *
+                            (block.timestamp - previousCheckpoint)) /
+                        sinceLast;
+                }
+                break;
+            } else {
+                /// @dev If passed weeks missed
+                if (sinceLast == 0 && nextWeek == previousCheckpoint) {
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev Store proportion of tokens for this week in the past
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute * (nextWeek - previousCheckpoint)) /
+                        sinceLast;
+                }
+            }
+            previousCheckpoint = nextWeek;
+            thisWeek = nextWeek;
+        }
+        emit CheckpointToken(block.timestamp, toDistribute);
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function claimEpoch(address to, uint epochNumber) public override {
+        _checkpointWhenReady();
+        _isEpochReady(epochNumber);
+        if (claimedEpochs[to][epochNumber] == true) {
+            revert CannotClaimTwice();
+        }
+        claimedEpochs[to][epochNumber] = true;
+
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            _startOfEpoch(epochNumber)
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
+        }
+
+        uint256 proportionalFees = calculateEpochFees(to, epochNumber);
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
+
+        emit EpochClaim(to, epochNumber, proportionalFees);
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function claimMany(address to, uint[] memory epochs) public {
+        for (uint i = 0; i < epochs.length; i++) {
```

Comment:
**Suggestion:** This can be optimised in a few ways:
```solidity
uint256 length = epochs.length; // this is more efficient as length isn't accessed each iteration
for (uint i = 0; i < length;) {
  // do stuff
  unchecked { // unchecked is more efficient as it doesn't check for overflows
    ++i; // notice here ++i is more efficient than i++
  }
}
```